### PR TITLE
[SPARK-44059] Add analyzer support of named arguments for built-in functions

### PR DESCRIPTION
**Pull request by learningchess2003**
### What changes were proposed in this pull request?
Add analyzer support for named function arguments. 

### Why are the changes needed?
Part of the project needed for general named function argument support.

### Does this PR introduce _any_ user-facing change?
We added support for named arguments for the ```CountMinSketchAgg``` and ```Mask``` SQL functions.

### How was this patch tested?
A new suite was added for this test called NamedArgumentFunctionSuite.

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -632,6 +632,12 @@
     ],
     "sqlState" : "22012"
   },
+  "DUPLICATE_ROUTINE_PARAMETER_ASSIGNMENT" : {
+    "message" : [
+      "Cannot invoke function <functionName> because it includes multiple arguments assigned to the same named parameter <parameterName>. At most one argument can be assigned to each parameter."
```
Comment (dtenedor) [inline core/src/main/resources/error/error-classes.json:637]:
```suggestion
      "Call to function <functionName> is invalid because it includes multiple named argument references with the same name <parameterName>."
```

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -1977,6 +1988,11 @@
     ],
     "sqlState" : "42614"
   },
+  "REQUIRED_PARAMETER_NOT_FOUND" : {
+    "message" : [
+      "Cannot invoke function <functionName> because the parameter named <parameterName> is required, but the function call did not supply a value."
```
Comment (dtenedor) [inline core/src/main/resources/error/error-classes.json:1993]:
```suggestion
      "Cannot invoke function <functionName> because the parameter named <parameterName> is required, but the function call did not supply a value. Please update the function call to supply an argument value (either positionally or by name) and retry the query again."
```

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -2187,6 +2208,11 @@
     ],
     "sqlState" : "42704"
   },
+  "UNRECOGNIZED_PARAMETER_NAME" : {
+    "message" : [
+      "Cannot invoke function <functionName> because the function call included a named argument reference for the argument named <argumentName>, but no function definition includes an argument with this name."
```
Comment (dtenedor) [inline core/src/main/resources/error/error-classes.json:2213]:
```suggestion
      "Cannot invoke function <functionName> because the function call included a named argument reference for the argument named <argumentName>, but this function does not include any signature containing an argument with this name."
```

File: sql/core/src/test/resources/sql-tests/inputs/mask-functions.sql [hunk 4]
```diff
@@ -5,6 +5,7 @@ SELECT mask('AbCD123-@$#', 'Q', 'q');
 SELECT mask('AbCD123-@$#', 'Q', 'q', 'd');
 SELECT mask('AbCD123-@$#', 'Q', 'q', 'd', 'o');
 SELECT mask('AbCD123-@$#', 'Qa', 'qa', 'da', 'oa');
+SELECT mask('AbCD123-@$#', lowerChar => 'q', upperChar => 'Q', otherChar => 'o', digitChar => 'd');
```
Comment (dtenedor) [inline sql/core/src/test/resources/sql-tests/inputs/mask-functions.sql:8]: can you add more .sql file test cases, including arguments of different types, error cases, etc.? Try to exercise all the corner cases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 5]
```diff
@@ -866,6 +866,61 @@ object FunctionRegistry {
...
+   * the expressions in the order specified by the object.
+   *
+   * It is here we resubstitute [[Unevaluable]] [[NamedArgumentExpression]]s with
+   * normal expressions.
+   *
+   * @param expressions The list of positional and named argument expressions
+   * @tparam T The actual expression class.
+   * @return positional argument list
+   */
+   def getRearrangedExpressions[T <: Expression : ClassTag](
```
Comment (dtenedor) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:881]: should we move this into `SupportsNamedArguments.scala`?

**Review by dtenedor — COMMENTED** [review]
Thanks for working on this! Took an initial review, left some suggestions for error messages and test cases for now.

File: sql/core/src/test/resources/sql-tests/inputs/mask-functions.sql [hunk 6]
```diff
@@ -55,4 +55,9 @@ SELECT mask(c1, replaceArg) from values('abcd-EFGH-8765-4321', 'a') as t(c1, rep
...
 SELECT mask(c1, replaceArg) from values('abcd-EFGH-8765-4321', 123) as t(c1, replaceArg);
 SELECT mask('abcd-EFGH-8765-4321', 'A', 'w', '');
-SELECT mask('abcd-EFGH-8765-4321', 'A', 'abc');
\ No newline at end of file
+SELECT mask('abcd-EFGH-8765-4321', 'A', 'abc');
+SELECT mask('AbCD123-@$#', lowerChar => 'q', upperChar => 'Q', otherChar => 'o', digitChar => 'd');
+SELECT mask(lowerChar => 'q', 'AbCD123-@$#', upperChar => 'Q', otherChar => 'o', digitChar => 'd');
+SELECT mask(lowerChar => 'q', upperChar => 'Q', otherChar => 'o', digitChar => 'd', str => 'AbCD123-@$#');
+SELECT mask('AbCD123-@$#', lowerChar => 'q', upperChar => 'Q', digitChar => 'd');
+SELECT mask(lowerChar => 'q', upperChar => 'Q', digitChar => 'd', str => 'AbCD123-@$#');
```
Comment (dtenedor) [inline sql/core/src/test/resources/sql-tests/inputs/mask-functions.sql:63]: Thanks for adding these tests! We'll want to add a lot more, including positive cases where the function arguments get successfully rearranged, as well as all the negative cases we can think of (to show the error messages), e.g.

* duplicate provided named arg
* provided named arg not found in the function signature
* provided named arg type is not coercible to the expected type
* provided named arg has value of `null`
* named arg appears before a positional arg
* expected named parameter does not have a corresponding provided argument

**Review by dtenedor — COMMENTED** [review]
Thanks again for working on this! I put in more review, thanks for adding the unit test coverage. I tried to think of more testing ideas.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        ensureNoNamedArgumentExpressions(expressions)
+      case _ =>
+        val moduleClassSymbol = targetModuleSymbol.asModule.moduleClass.asClass
+        moduleClassSymbol.baseClasses.contains(parentSymbol) match {
+          case true =>
+            if (currentMirror.runtimeClass(moduleClassSymbol).getEnclosingClass != null) {
+              throw QueryCompilationErrors.cannotObtainCompanionObjectInstance(functionName)
+            }
+            val instance = currentMirror.reflectModule(targetModuleSymbol.asModule)
+              .instance.asInstanceOf[SupportsNamedArguments]
```
[the remark this answers is not in the document]
Comment (learningchess2003) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:131]: @dtenedor  Ah, you see. This is reflection magic at work. You might've noticed in the above line that I do a check where we use the ```baseClasses``` method and check if it contains ```parentSymbol```. This is where the check for whether or not the provided companion object extends the ```SupportsNamedArgument``` trait. Therefore, the check here will be redundant. Second, I mentioned previously that if the companion object is not top level, then we cannot obtain the instance. Therefore, it would be a good idea to only try to obtain the instance _after_ we are sure that we absolutely require it which is when we need to call the ```functionSignatures``` method. In the case that we want the companion object, I do a check to see if the module class which represents the companion object is the member of an enclosing class. This is the check to see if the companion object is top level object to ensure it is obtainable via reflection without providing an instance of an enclosing class. If it is not, then that means it was the intention of the developer that for this function expression to support named arguments, but reflection is not powerful enough to get the function signature information we desire. By this point, we can then directly obtain the instance, know that it is a ```SupportsNamedArguments``` instance, and call ```functionSignatures``` directly since the validation had already been previously performed.

[the remark this answers is not in the document]
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:131]: OK, sounds good, please paraphrase that into a code comment here :)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+abstract class SupportsNamedArguments {
+  def functionSignatures: Seq[FunctionSignature]
+}
+
+object SupportsNamedArguments {
+  final def rearrange(functionSignature: FunctionSignature,
+                      args: Seq[Expression],
+                      functionName: String): Seq[Expression] = {
+    val parameters = functionSignature.parameters
+    if (!args.exists(_.isInstanceOf[NamedArgumentExpression])) {
```
[the remark this answers is not in the document]
Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:42]: Or you can return here so that the rest of else doesn't indent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      args
+    } else {
+      val firstNamedArgIdx = args.indexWhere(_.isInstanceOf[NamedArgumentExpression])
+      val (positionalArgs, namedArgs) = args.splitAt(firstNamedArgIdx)
+      val namedParameters = parameters.drop(positionalArgs.size)
+
+      // Performing some checking to ensure valid argument list
+      val allNamedParameterNames = parameters.map(_.name)
+      val parameterNamesSet = allNamedParameterNames.toSet
+      val assignedParameterSet = collection.mutable.Set[Expression]() ++
```
Comment (anchovYu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:52]: Is the intention here to convert the immutable to mutable set? Could you use something like `collection.mutable.Set[Expression](..: _*)`?

File: core/src/main/resources/error/error-classes.json [hunk 10]
```diff
@@ -632,6 +632,12 @@
     ],
     "sqlState" : "22012"
   },
+  "DUPLICATE_ROUTINE_PARAMETER_ASSIGNMENT" : {
+    "message" : [
+      "Call to function <functionName> is invalid because it includes multiple named argument references with the same name <parameterName>."
+    ],
+    "sqlState" : "42704K"
```
[the remark this answers is not in the document]
Comment (anchovYu) [reply core/src/main/resources/error/error-classes.json:639]: It also includes the case that reassigning values with named argument, on a previously specified positional argument.

File: core/src/main/resources/error/error-classes.json [hunk 11]
```diff
@@ -2152,6 +2168,11 @@
     ],
     "sqlState" : "42K09"
   },
+  "UNEXPECTED_POSITIONAL_ARGUMENT" : {
```
Comment (anchovYu) [inline core/src/main/resources/error/error-classes.json:2171]: Good error message!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 12]
```diff
@@ -50,6 +50,59 @@ import org.apache.spark.sql.types._
  */
 private[sql] object QueryCompilationErrors extends QueryErrorsBase {
 
+  def cannotObtainCompanionObjectInstance(functionName: String): Throwable = {
+    SparkException.internalError(s"Cannot obtain companion object for " +
+      s"function expression: $functionName. Companion must be top-level object.")
+  }
+
+  def multipleFunctionSignatures(functionName: String): Throwable = {
+    SparkException.internalError(s"Function $functionName cannot have multiple method signatures.")
```
[the remark this answers is not in the document]
Comment (anchovYu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:59]: The error message is not about ambiguous signatures but instead intended for we don't allow having multiple signatures now. But it's no harm to print out more info in the error messages.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/NamedArgumentFunctionSuite.scala [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+  override def nullable: Boolean = false
+  override def eval(input: InternalRow): Any = None
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = null
+  override def dataType: DataType = null
+  override def children: Seq[Expression] = Nil
+  override protected def withNewChildrenInternal(
+    newChildren: IndexedSeq[Expression]): Expression = null
+}
+
+class NamedArgumentFunctionSuite extends AnalysisTest {
```
Comment (anchovYu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/NamedArgumentFunctionSuite.scala:76]: We should include successful test cases as well. We may consider to gather tests for all functions extending the trait here.
We can increase the test scope a bit by include the transition from the unresolved function to a fully resolved expression. e.g. 
```
plan = UnresolvedFunction(..)
expected = ..
assert(getAnalyzer.execute(plan) == expected)
```

**Comment by learningchess2003** [comment]
@dtenedor @anchovYu Thanks for the reviews! Trying to address the comments to the best of my understanding. I did have one question regarding ```rearrange``` and which function should be overrideable (be it not recommended). From my understanding, I would've thought that it be the ```rearrange``` function rather than the ```getRearrangedExpressions``` function since the latter performs some mandatory reflection procedures which really shouldn't be overriden. In which case, it would just be the ```rearrange``` function then. If this doesn't sound right, please let me know! 

I changed the code according to this understanding. So it might look different than what you envisioned.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+trait NamedArgumentType
+
+case class FixedArgumentType(dataType: AbstractDataType) extends NamedArgumentType
+
+case class NamedArgument(name: String,
+                         dataType: NamedArgumentType,
+                         default: Option[Expression] = None)
+case class FunctionSignature(parameters: Seq[NamedArgument])
+abstract class SupportsNamedArguments {
+  def functionSignatures: Seq[FunctionSignature]
```
[the remark this answers is not in the document]
Comment (learningchess2003) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:60]: I don't think this would work directly because it is very hard to provide the function name to the error message. We can add an extra field ```def functionName: String``` which can be overriden that does provide it if you want.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 15]
```diff
@@ -50,6 +50,59 @@ import org.apache.spark.sql.types._
  */
 private[sql] object QueryCompilationErrors extends QueryErrorsBase {
 
+  def cannotObtainCompanionObjectInstance(functionName: String): Throwable = {
+    SparkException.internalError(s"Cannot obtain companion object for " +
```
[the remark this answers is not in the document]
Comment (learningchess2003) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:54]: This comment is used in the context that the user already tried to have their companion object extend SupportsNamedArguments, but fail because it is not a top level object. In this case, the message I gave should be pretty concise and explain why it didn't work pretty well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+   * the expressions in the order specified by the object.
+   *
+   * It is here we resubstitute [[Unevaluable]] [[NamedArgumentExpression]]s with
+   * normal expressions.
+   *
+   * @param expressions The list of positional and named argument expressions
+   * @tparam T The actual expression class.
+   * @return positional argument list
+   */
+  final def getRearrangedExpressions[T <: Expression : ClassTag](
```
[the remark this answers is not in the document]
Comment (learningchess2003) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala:96]: Yeah, that sounds good.

File: core/src/main/resources/error/error-classes.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (learningchess2003) [reply core/src/main/resources/error/error-classes.json:639]: That is true. We can modify the error message a little to include that.

[the remark this answers is not in the document]
Comment (learningchess2003) [reply core/src/main/resources/error/error-classes.json:639]: Oh yeah, also the K should be there. Yaohua also added it in his PR for read files.

File: core/src/main/resources/error/error-classes.json [hunk 17]
```diff
@@ -1659,6 +1665,11 @@
       "Not allowed to implement multiple UDF interfaces, UDF class <className>."
     ]
   },
+  "NAMED_ARGUMENTS_NOT_SUPPORTED" : {
```
[the remark this answers is not in the document]
Comment (learningchess2003) [reply core/src/main/resources/error/error-classes.json:1668]: Not all the error classes have sqlStates. This is just a not supported error, so there isn't any SQL error standard to reference.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (learningchess2003) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/SupportsNamedArguments.scala]: Actually, upon some further thought, I realized that this is a bad idea. I don't think we should do something like this because that means if we change the function name which the case class is binding to, we need to change it both in the FunctionRegistry and here. This will only be more annoying to maintain and easy to miss. I don't think this will be a good idea.

**Review by anchovYu — COMMENTED** [review]
Thanks for this effort! Good work!
Could you also link the doc and your last related PR in the description? That can provide context for other reviewers.

**Comment by learningchess2003** [comment]
Due to some problems with rebasing and merge conflicts with the upstream branch, I'm migrating to PR #2 which is the same thing. Will continue to refer to this PR for comments regarding testing.

**Comment by learningchess2003** [comment]
Actually, will keep this PR since it doesn't look like it will be affected.
